packages/sandcastle/src/App.tsx

```diff
@@ -241,11 +247,46 @@ function App() {
           committedCode: action.code ?? state.code,
           committedHtml: action.html ?? state.html,
           runNumber: state.runNumber + 1,
+          dirty: false,
+        };
+      }
+      case "resetDirty": {
+        return {
+          ...state,
+          dirty: false,
         };
       }
     }
   }, initialState);
 
+  useEffect(() => {
+    const host = window.location.host;
+    let envString = "";
+    if (host.includes("localhost") && host !== "localhost:8080") {
+      // this helps differentiate tabs for local sandcastle development or other testing
+      envString = `${host.replace("localhost:", "")} `;
+    }
+
+    const baseTitle = "Cesium Sandcastle";
+    const dirtyIndicator = codeState.dirty ? "*" : "";
+    if (title !== "") {
+      document.title = `${envString}${title}${dirtyIndicator} | ${baseTitle}`;
+    } else {
+      document.title = `${envString}${baseTitle}`;
+    }
+  }, [title, codeState.dirty]);
+
+  const confirmLeave = useCallback(() => {
+    if (!codeState.dirty) {
+      return true;
+    }
+
+    /* eslint-disable-next-line no-alert */
+    return window.confirm(
+      "You have unsaved changes. Are you sure you want to navigate away from this demo?",
+    );
```

Comment:
This is totally up-to-par with what we had before, but what do you think of propagating the current sandcastle to local storage or similar?

The goal would to be to persist state, like what happens when you reload a GitHub tab with an in-progress comment. Or maybe a slightly different workflow, like what happens if you leave and come back to https://geojson.io/.